radix-engine/src/utils/native_blueprint_call_validator.rs

```diff
@@ -0,0 +1,460 @@
+use radix_engine_interface::blueprints::account::*;
+use radix_engine_interface::blueprints::consensus_manager::*;
+use radix_engine_interface::blueprints::identity::*;
+use radix_engine_interface::blueprints::resource::*;
+use radix_engine_interface::schema::BlueprintSchema;
+use sbor::validate_payload_against_schema;
+use sbor::LocalTypeIndex;
+use sbor::Schema;
+use transaction::prelude::InstructionV1;
+
+pub fn validate_call_arguments_to_native_components(
+    instructions: &[InstructionV1],
+) -> Result<(), ValidationError> {
+    let mut package_definitions = LazyPackageDefinitions::new();
+
+    for (index, instruction) in instructions.iter().enumerate() {
+        let (invocation, args) = match instruction {
+            InstructionV1::CallFunction {
+                package_address,
+                blueprint_name,
+                function_name,
+                args,
+            } => (
+                Invocation::Function(
+                    *package_address,
+                    blueprint_name.to_owned(),
+                    function_name.to_owned(),
+                ),
+                args,
+            ),
+            InstructionV1::CallMethod {
+                address,
+                method_name,
+                args,
+            } => (
+                Invocation::Method(*address, ObjectModuleId::Main, method_name.to_owned()),
+                args,
+            ),
+            InstructionV1::CallMetadataMethod {
+                address,
+                method_name,
+                args,
+            } => (
+                Invocation::Method(*address, ObjectModuleId::Metadata, method_name.to_owned()),
+                args,
+            ),
+            InstructionV1::CallRoyaltyMethod {
+                address,
+                method_name,
+                args,
+            } => (
+                Invocation::Method(*address, ObjectModuleId::Royalty, method_name.to_owned()),
+                args,
+            ),
+            InstructionV1::CallAccessRulesMethod {
+                address,
+                method_name,
+                args,
+            } => (
+                Invocation::Method(
+                    *address,
+                    ObjectModuleId::AccessRules,
+                    method_name.to_owned(),
+                ),
+                args,
+            ),
+            _ => continue,
+        };
+
+        let schema = get_arguments_schema(invocation, &mut package_definitions, index)?;
+        if let Some((local_type_index, schema)) = schema {
+            validate_payload_against_schema::<ManifestCustomExtension, _>(
+                &manifest_encode(&args).unwrap(),
+                schema,
+                local_type_index,
+                &(),
+            )
+            .map_err(|error| ValidationError::SchemaValidationError(index, format!("{:?}", error)))
+        } else {
+            Ok(())
+        }?;
+    }
+
+    Ok(())
+}
+
+macro_rules! define_lazy_package_definitions {
```

Comment:
Out of interest, why don't we use lazy static here?

It would be helpful to have this statically defined in the node :)

Reply:
It can definitely use lazy static here. My thought when I did this was that we would only be using it here, thus having the schemas as globals didn't make a lot of sesne. However, as you mentioned in your other comment the node wants to use this, and so does the toolkit. So, I will use lazy static here.